src/gs-background-tile.c

```diff
@@ -0,0 +1,276 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include <glib/gi18n.h>
+#include <gtk/gtk.h>
+
+#include "gs-background-tile.h"
+#include "gs-star-widget.h"
+#include "gs-common.h"
+
+struct _GsBackgroundTile
+{
+	GsAppTile	 parent_instance;
+
+	GsApp		*app;
+	GtkWidget	*name_label;
+	GtkWidget	*summary_label;
+	GtkWidget	*image;
+	GtkWidget	*image_box;
+	GtkWidget	*installed_icon;
+	GtkWidget	*price;
+	GtkWidget	*stack;
+};
+
+G_DEFINE_TYPE (GsBackgroundTile, gs_background_tile, GS_TYPE_APP_TILE)
+
+static GsApp *
+gs_background_tile_get_app (GsAppTile *tile)
+{
+	return GS_BACKGROUND_TILE (tile)->app;
+}
+
+static void
+gs_background_tile_setup_colors_bg (GsBackgroundTile *tile)
+{
+	GsApp *app = tile->app;
+	const guint num_colors = 4;
+	const GPtrArray *colors = gs_app_get_key_colors (app);
+	const gchar *gradients[num_colors];
+	g_autofree gchar *css = NULL;
+	gint degrees = 45;
+
+	if (colors->len == 0)
+		return;
+
+	/* apply one color gradient for each color from the cornors in the
+	 * following fashion:
+	 * -------  -------  -------  -------
+	 * -     -  - /   -  - / \ -  - / \ -
+	 * - \   -  - \   -  - \   -  - \ / -
+	 * -------  -------  -------  -------
+	 */
+	for (guint i = 0; i < num_colors; ++i) {
+		GdkRGBA *rgba = g_ptr_array_index (colors, i % colors->len);
+		guint red, green, blue;
+
+		/* we use the values in the printf below instead of using the
+		 * convenient gdk_rgba_to_string to avoid extra allocations as
+		 * this function can be used many times */
+		red = (guint) (255.0 * rgba->red);
+		green = (guint) (255.0 * rgba->green);
+		blue = (guint) (255.0 * rgba->blue);
+
+		/* the gradient will go from an solid color to the same color with more
+		 * transparency */
+		gradients[i] = g_strdup_printf ("linear-gradient(%udeg, "
+						"rgba(%u, %u, %u, 1.0), "
+						"rgba(%u, %u, %u, .2) 40%%)",
+						degrees + i * 90,
+						red, green, blue,
+						red, green, blue);
+	}
+
+	css = g_strconcat ("background: ",
+			   gradients[0], ",",
+			   gradients[1], ",",
+			   gradients[2], ",",
+			   gradients[3], ";",
+			   NULL);
+	gs_utils_widget_set_css (GTK_WIDGET (tile->image_box), css);
+
+	for (guint i = 0; i < num_colors; ++i)
+		g_free (gradients[i]);
+}
+
+static void
+gs_background_tile_update_price (GsBackgroundTile *tile)
+{
+	const GsPrice *price = NULL;
+
+	if (gs_app_is_installed (tile->app)) {
+		gtk_widget_hide (tile->price);
+		gtk_widget_show (tile->installed_icon);
+		return;
+	}
+
+	price = gs_app_get_price (tile->app);
+	if (price == NULL || (gs_price_get_amount (price) == (gdouble) 0))
+		gtk_label_set_label (GTK_LABEL (tile->price), _("FREE"));
```

Comment:
I'd add a translator's comment to keep this upper case, or even use a simple string here and then use e.g. `g_strup()` to make it upper case.

Reply:
I wonder if it's worth that, and whether that will drop some needed context for other languages. But maybe not.